### PR TITLE
Separate code deposit cost in semantic tests

### LINKED ISSUE #13160: [soltest] Distinguish code size-induced cost from runtime gas cost in tests.
Currently, the semantics test expectations contain full transaction gas costs without distinguishing between cost due to code size vs runtime gas cost. While this is a bit tricky to distinguish properly, it'd be good to try and separate them to make the gas values more meaningful.

We could probably have ``EVMHost`` (``test/EVMHost.*``) track the cost associated with code size on each and every creation during a transaction (just accummulating the costs of multiple contract creations), then have the ``SemanticTest``s subtract that from the runtime gas costs, while reporting the code size separately. (We could also try to account for the cost of the bytecode size in the payload of the creation transactions, but it'd be an improvement even without that.)

Then we'd have to adjust ``scripts/gas_diff_stats.py`` to list code size changes and runtime gas changes separately (and also make scripts/gas_diff_stats.py itself executable, it's mildly annoying that it's not).

### LINKED ISSUE #14477: Split code size cost from runtime gas cost in semantics tests.
Our semantics test framework runs EVM execution through evmone and collects runtime gas statistics.

While the test expectations distinguish between the cost of the "constructor" and of calling methods, our test cases regularly themselves deploy contracts, yielding a mixture of runtime cost and code size as result.

While there is some logic to that (contracts deploying contracts, of course, is a real world use case), it makes evaluating optimizer changes challenging.

So ideally, we should split the gas costs in our semantics test expectations into code size cost and deploy cost by tracking any contract creation, subtracting the code cost from the runtime gas cost and listing it separately.

### PR DESCRIPTION
**Pull request by cameel**
Fixes #14477.
Fixes #13160.
~Depends on #14842.~ Merged.

### CODE REVIEW

File: test/scripts/test_gas_diff_stats.py [hunk 1]
```diff
@@ -85,3 +85,25 @@ def test_collect_statistics_should_ignore_unchanged_costs(self):
...
+        """).splitlines()
+
+        self.assertEqual(collect_statistics(diff_output), (
+            1,         # -irOptimized
+            300,       # -legacyOptimized
+            900000020, # -legacy
+            54000,     # +irOptimized
+            87000000,  # +legacyOptimized
+            600000,    # +legacy
+        ))
```
Comment (cameel) [inline test/scripts/test_gas_diff_stats.py:109]: Note: The totals reported reported by `gas_diff_stats.py` do include the code deposit costs.

File: test/libsolidity/semanticTests/abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol [hunk 2]
```diff
@@ -30,6 +30,9 @@ contract C is B {
...
 // test() -> 77
-// gas irOptimized: 110348
-// gas legacy: 151866
-// gas legacyOptimized: 110373
+// gas irOptimized: 55148
+// gas irOptimized code: 55200
+// gas legacy: 57266
+// gas legacy code: 94600
+// gas legacyOptimized: 55173
+// gas legacyOptimized code: 55200
```
Comment (cameel) [inline test/libsolidity/semanticTests/abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol:38]: I went with this format for simplicity, but I don't like how verbose it is. Comparing gas changes in PR diffs was already tedious even without the new entries.

I think that something along these lines would be more readable, especially when differences are small and when there's highlighting of differences within the line (which seems not enabled in github comments):
```diff
-// gas: 55148+55200 ir (optimized), 7000+94600 legacy, 55173+55200 legacy (optimized)
+// gas: 55148+55200 ir (optimized), 57266+94600 legacy, 55173+55200 legacy (optimized)
```

And even with large differences, the misalignment actually makes it easy to spot such, more important, lines:
```diff
-// gas: 110348 ir (optimized), 151866 legacy, 110373 legacy (optimized)
+// gas: 55148+55200 ir (optimized), 57266+94600 legacy, 55173+55200 legacy (optimized)
```

In the end I did not have time to do that though. I might try later when I have some spare time between bigger tasks.

File: test/libsolidity/SemanticTest.cpp [hunk 3]
```diff
@@ -579,17 +584,23 @@ bool SemanticTest::checkGasCostExpectation(TestFunctionCall& io_test, bool _comp
 		!m_enforceGasCost ||
 		m_gasUsed < m_enforceGasCostMinValue ||
```
Comment (cameel) [inline test/libsolidity/SemanticTest.cpp:585]: Note that gas enforcement threshold is still based on the total and does not exclude code deposit cost. I'm not really sure which is better, but neither seems wrong either so I kept it at that.

**Comment by ekpyron** [comment]
I'm fine with the format as is in the PR - so I'd say let's just try to get this merged quickly as well, s.t. we can start using the data - and we can still at a later point change it further, if we ever see the need!

**Comment by cameel** [comment]
Great. It can go in as soon as #14842 is merged. I just need to rebase and regenerate gas costs on top of `develop`.

**Comment by ekpyron** [comment]
> Great. It can go in as soon as #14842 is merged. I just need to rebase and regenerate gas costs on top of `develop`.

So, do you want to rebase it now :-)? I'd myself also squash it, but as you like.

**Comment by cameel** [comment]
Rebased. I'd rather leave commits separated though.

**Review by ekpyron — APPROVED** [review]

**Comment by cameel** [comment]
WTF. #14875 is merged and I just rebased on `develop`. Why is `t_ems_ext_hardhat` still failing?

I mean, it's not required do we can still merge, but still, it should be passing now...

**Comment by cameel** [comment]
> ```
> /bin/sh: 1: cargo: not found
> ```

Ah, I guess this is a separate issue that we don't have a fix for yet? I remember switching to the rust/node image being mentioned on the channel so it's probably that.

**Pull request merged**